### PR TITLE
Add system-characteristics 'information-type' constraints

### PR DESCRIPTION
**Pull request by Gabeblis**
# Committer Notes

{Please provide a description of what this PR accomplishes. Be sure to reference any issues addressed. If the PR is a work-in-progress submitted for early review, please submit the PR as a draft PR using the "Draft pull request" dropdown.}

### Description 
Added system-characteristics 'information-type' constraints and tests for each of the constraints. These should be added as they are part of the effort write all of the constraints from the constraint tracker.  

### Changes 
- Added constraints information-type-has-confidentiality-impact, information-type-has-integrity-impact, information-type-has-availability-impact to fedramp-external-constraints.xml.  
- Added pass and fail yaml tests for all of the above constraints.  
- Edited ssp-all-INVALID.xml to ensure all constraints fail correctly.  

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/GSA/fedramp-automation/blob/master/CONTRIBUTING.md) guidance?
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/GSA/fedramp-automation/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] If applicable, have all [FedRAMP Documents Related to OSCAL Adoption](https://github.com/GSA/fedramp-automation) affected by the changes in this issue have been updated.?
- ~If applicable, does this PR reference the issue it addresses and explain how it addresses the issue?~

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### CODE REVIEW

**Review by aj-stein-gsa — CHANGES_REQUESTED** [review]
Does this also have overlap with #610? Seems so.

**Comment by Gabeblis** [comment]
> Does this also have overlap with #610? Seems so.

These constraints should all exist. The only overlap is the issue that we discussed yesterday with Dave regarding allowed-values constraints and expect constraints sharing the same ssp-all-INVALID file.

**Comment by aj-stein-gsa** [comment]
> > Does this also have overlap with #610? Seems so.
> 
> These constraints should all exist. The only overlap is the issue that we discussed yesterday with Dave regarding allowed-values constraints and expect constraints sharing the same ssp-all-INVALID file.

OK disregard, I can see what you mean. I am still going to need to wrap up some docs pertaining to the doc gaps for this and #610 tracked in https://github.com/GSA/automate.fedramp.gov/issues/55. Sorry for the delayed reply.

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 1]
```diff
@@ -164,6 +164,18 @@
                 <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/4-ssp-template-to-oscal-mapping/#system-sensitivity-level"/>
                 <message>A FedRAMP SSP SHOULD define its FIPS-199 security sensitivity level to match the highest security impact level for the system's confidentiality, integrity, and availability objectives.</message>
             </expect>
+            <expect id="information-type-has-confidentiality-impact" target="system-characteristics/system-information/information-type" test="confidentiality-impact" level="ERROR">
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/4-ssp-template-to-oscal-mapping/#system-information-and-information-types"/>
+                <message>An OSCAL SSP information type MUST have a confidentiality impact.</message>
```
Comment (DimitriZhurkin) [inline src/validations/constraints/fedramp-external-constraints.xml]: The message should probably start with `A FedRAMP SSP...`

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 2]
```diff
@@ -164,6 +164,18 @@
...
+            </expect>
+            <expect id="information-type-has-integrity-impact" target="system-characteristics/system-information/information-type" test="integrity-impact" level="ERROR">
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/4-ssp-template-to-oscal-mapping/#system-information-and-information-types"/>
+                <message>An OSCAL SSP information type MUST have an integrity impact.</message>
```
Comment (DimitriZhurkin) [inline src/validations/constraints/fedramp-external-constraints.xml]: Ditto message.

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 3]
```diff
@@ -164,6 +164,18 @@
...
+            </expect>
+            <expect id="information-type-has-availability-impact" target="system-characteristics/system-information/information-type" test="availability-impact" level="ERROR">
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/4-ssp-template-to-oscal-mapping/#system-information-and-information-types"/>
+                <message>An OSCAL SSP information type MUST have an availability impact.</message>
```
Comment (DimitriZhurkin) [inline src/validations/constraints/fedramp-external-constraints.xml]: Ditto message.

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 4]
```diff
@@ -164,6 +164,18 @@
...
                 <message>A FedRAMP SSP SHOULD define its FIPS-199 security sensitivity level to match the highest security impact level for the system's confidentiality, integrity, and availability objectives.</message>
             </expect>
+            <expect id="information-type-has-confidentiality-impact" target="system-characteristics/system-information/information-type" test="confidentiality-impact" level="ERROR">
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/4-ssp-template-to-oscal-mapping/#system-information-and-information-types"/>
+                <message>An OSCAL SSP information type MUST have a confidentiality impact.</message>
+            </expect>
+            <expect id="information-type-has-integrity-impact" target="system-characteristics/system-information/information-type" test="integrity-impact" level="ERROR">
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/4-ssp-template-to-oscal-mapping/#system-information-and-information-types"/>
```
Comment (DimitriZhurkin) [inline src/validations/constraints/fedramp-external-constraints.xml:172]: Ditto `help-url`.

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 5]
```diff
@@ -164,6 +164,18 @@
...
+                <message>An OSCAL SSP information type MUST have an integrity impact.</message>
+            </expect>
+            <expect id="information-type-has-availability-impact" target="system-characteristics/system-information/information-type" test="availability-impact" level="ERROR">
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/4-ssp-template-to-oscal-mapping/#system-information-and-information-types"/>
```
Comment (DimitriZhurkin) [inline src/validations/constraints/fedramp-external-constraints.xml:176]: Ditto `help-url`.

File: src/validations/constraints/fedramp-external-constraints.xml [hunk 6]
```diff
@@ -164,6 +164,18 @@
                 <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/4-ssp-template-to-oscal-mapping/#system-sensitivity-level"/>
                 <message>A FedRAMP SSP SHOULD define its FIPS-199 security sensitivity level to match the highest security impact level for the system's confidentiality, integrity, and availability objectives.</message>
             </expect>
+            <expect id="information-type-has-confidentiality-impact" target="system-characteristics/system-information/information-type" test="confidentiality-impact" level="ERROR">
+                <prop namespace="https://docs.oasis-open.org/sarif/sarif/v2.1.0" name="help-url" value="https://automate.fedramp.gov/documentation/ssp/4-ssp-template-to-oscal-mapping/#system-information-and-information-types"/>
```
Comment (DimitriZhurkin) [inline src/validations/constraints/fedramp-external-constraints.xml:168]: In the `help-url` URL, the `#system-information-and-information-types` anchor  didn't work for me.

Is it work in progress?

File: src/validations/constraints/unit-tests/information-type-has-availability-impact-FAIL.yaml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+test-case:
+  name: Negative Test for information-type-has-availability-impact
+  description: Test that a SSP system-characteristics system-information information-type element does not have an availability-impact element.
```
Comment (DimitriZhurkin) [inline src/validations/constraints/unit-tests/information-type-has-availability-impact-FAIL.yaml:3]: `a SSP` should be `an SSP`.

File: src/validations/constraints/unit-tests/information-type-has-availability-impact-PASS.yaml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+test-case:
+  name: Positive Test for information-type-has-availability-impact
+  description: Test that a SSP system-characteristics system-information information-type element has an availability-impact element.
```
Comment (DimitriZhurkin) [inline src/validations/constraints/unit-tests/information-type-has-availability-impact-PASS.yaml:3]: Ditto `a SSP`.

File: src/validations/constraints/unit-tests/information-type-has-confidentiality-impact-FAIL.yaml [hunk 9]
```diff
@@ -0,0 +1,7 @@
+test-case:
+  name: Negative Test for information-type-has-confidentiality-impact
+  description: Test that a SSP system-characteristics system-information information-type element does not have a confidentiality-impact element.
```
Comment (DimitriZhurkin) [inline src/validations/constraints/unit-tests/information-type-has-confidentiality-impact-FAIL.yaml:3]: Ditto `a SSP`.

File: src/validations/constraints/unit-tests/information-type-has-confidentiality-impact-PASS.yaml [hunk 10]
```diff
@@ -0,0 +1,7 @@
+test-case:
+  name: Positive Test for information-type-has-confidentiality-impact
+  description: Test that a SSP system-characteristics system-information information-type element has a confidentiality-impact element.
```
Comment (DimitriZhurkin) [inline src/validations/constraints/unit-tests/information-type-has-confidentiality-impact-PASS.yaml:3]: Ditto `a SSP`.

File: src/validations/constraints/unit-tests/information-type-has-integrity-impact-FAIL.yaml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+test-case:
+  name: Negative Test for information-type-has-integrity-impact
+  description: Test that a SSP system-characteristics system-information information-type element does not have an integrity-impact element.
```
Comment (DimitriZhurkin) [inline src/validations/constraints/unit-tests/information-type-has-integrity-impact-FAIL.yaml:3]: Ditto `a SSP`.

File: src/validations/constraints/unit-tests/information-type-has-integrity-impact-PASS.yaml [hunk 12]
```diff
@@ -0,0 +1,7 @@
+test-case:
+  name: Positive Test for information-type-has-integrity-impact
+  description: Test that a SSP system-characteristics system-information information-type element has an integrity-impact element.
```
Comment (DimitriZhurkin) [inline src/validations/constraints/unit-tests/information-type-has-integrity-impact-PASS.yaml:3]: Ditto `a SSP`.

**Review by DimitriZhurkin — CHANGES_REQUESTED** [review]

File: src/validations/constraints/fedramp-external-constraints.xml (hunk 1 above)
Comment (Gabeblis) [reply src/validations/constraints/fedramp-external-constraints.xml]: Thanks. Resolved: [1311a8d](https://github.com/GSA/fedramp-automation/pull/685/commits/1311a8de68b0f4d6ff507640fdd2e041cd912557)

File: src/validations/constraints/fedramp-external-constraints.xml (hunk 2 above)
Comment (Gabeblis) [reply src/validations/constraints/fedramp-external-constraints.xml]: Resolved: [1311a8d](https://github.com/GSA/fedramp-automation/pull/685/commits/1311a8de68b0f4d6ff507640fdd2e041cd912557)

File: src/validations/constraints/fedramp-external-constraints.xml (hunk 3 above)
Comment (Gabeblis) [reply src/validations/constraints/fedramp-external-constraints.xml]: Resolved: [1311a8d](https://github.com/GSA/fedramp-automation/pull/685/commits/1311a8de68b0f4d6ff507640fdd2e041cd912557)

File: src/validations/constraints/unit-tests/information-type-has-availability-impact-FAIL.yaml (hunk 7 above)
Comment (Gabeblis) [reply src/validations/constraints/unit-tests/information-type-has-availability-impact-FAIL.yaml]: Thanks! Resolved: [d5768dc](https://github.com/GSA/fedramp-automation/pull/685/commits/d5768dcd9df6f71e97829806125b7cff449faa20)

File: src/validations/constraints/unit-tests/information-type-has-availability-impact-PASS.yaml (hunk 8 above)
Comment (Gabeblis) [reply src/validations/constraints/unit-tests/information-type-has-availability-impact-PASS.yaml]: Resolved: [d5768dc](https://github.com/GSA/fedramp-automation/pull/685/commits/d5768dcd9df6f71e97829806125b7cff449faa20)

File: src/validations/constraints/unit-tests/information-type-has-confidentiality-impact-FAIL.yaml (hunk 9 above)
Comment (Gabeblis) [reply src/validations/constraints/unit-tests/information-type-has-confidentiality-impact-FAIL.yaml]: Resolved: [d5768dc](https://github.com/GSA/fedramp-automation/pull/685/commits/d5768dcd9df6f71e97829806125b7cff449faa20)

File: src/validations/constraints/unit-tests/information-type-has-confidentiality-impact-PASS.yaml (hunk 10 above)
Comment (Gabeblis) [reply src/validations/constraints/unit-tests/information-type-has-confidentiality-impact-PASS.yaml]: Resolved: [d5768dc](https://github.com/GSA/fedramp-automation/pull/685/commits/d5768dcd9df6f71e97829806125b7cff449faa20)

File: src/validations/constraints/unit-tests/information-type-has-integrity-impact-FAIL.yaml (hunk 11 above)
Comment (Gabeblis) [reply src/validations/constraints/unit-tests/information-type-has-integrity-impact-FAIL.yaml]: Resolved: [d5768dc](https://github.com/GSA/fedramp-automation/pull/685/commits/d5768dcd9df6f71e97829806125b7cff449faa20)

File: src/validations/constraints/unit-tests/information-type-has-integrity-impact-PASS.yaml (hunk 12 above)
Comment (Gabeblis) [reply src/validations/constraints/unit-tests/information-type-has-integrity-impact-PASS.yaml]: Resolved: [d5768dc](https://github.com/GSA/fedramp-automation/pull/685/commits/d5768dcd9df6f71e97829806125b7cff449faa20)

File: src/validations/constraints/fedramp-external-constraints.xml (hunk 4 above)
Comment (Gabeblis) [reply src/validations/constraints/fedramp-external-constraints.xml:172]: Thanks for catching that. Yes, the [documentation](https://github.com/GSA/automate.fedramp.gov/pull/87) has been merged into develop in automate.fedramp. The develop branch hasn't been merged into main with this updated content just yet, so you will not see in on the website until that happens.

File: src/validations/constraints/fedramp-external-constraints.xml (hunk 5 above)
Comment (Gabeblis) [reply src/validations/constraints/fedramp-external-constraints.xml:176]: Thanks for catching that. Yes, the [documentation](https://github.com/GSA/automate.fedramp.gov/pull/87) has been merged into develop in automate.fedramp. The develop branch hasn't been merged into main with this updated content just yet, so you will not see in on the website until that happens.

File: src/validations/constraints/fedramp-external-constraints.xml (hunk 6 above)
Comment (Gabeblis) [reply src/validations/constraints/fedramp-external-constraints.xml:168]: Thanks for catching that. Yes, the [documentation](https://github.com/GSA/automate.fedramp.gov/pull/87) has been merged into develop in automate.fedramp. The develop branch hasn't been merged into main with this updated content just yet, so you will not see in on the website until that happens.

**Review by DimitriZhurkin — APPROVED** [review]

**Review by wandmagic — APPROVED** [review]

**Review by aj-stein-gsa — APPROVED** [review]

**Pull request merged**